File: BOJ/Swift/2470.swift
import Foundation

/*
 산성 용액의 특성값: 1...1000000000
 알칼리성 용액의 특성값: -1...-1000000000
 같은 양의 두 용액을 혼합한 용액의 특성값 -> 혼합에 사용된 각 용액의 특성값의 합
 서로 다른 용액을 혼합해 특성값이 "0"에 가장 가까운 용액을 만들어내는 두 용액을 찾아라.
 */

let n = Int(readLine()!)!
var array = readLine()!.split(separator: " ").map{ Int(String($0))! }.sorted()
var sum = 2000000001
var left = 0
var right = n-1
var ans1 = 0
var ans2 = 0

while left < right {
    if sum > abs(array[left] + array[right]) {
        sum = abs(array[left] + array[right])
        ans1 = left
        ans2 = right
    }
    if array[left] + array[right] > 0 {
        right -= 1
    }else {
        left += 1
    }
}
print("\(array[ans1]) \(array[ans2])")

File: BOJ/cpp/2470.cpp
//두 포인터
#include <iostream>
#include <algorithm>
#include <cmath>
using namespace std;
int n, a[100001];

void input(){
    cin >> n;
    for(int i=0; i<n; i++) cin >> a[i];
}

void solution(){
    sort(a, a+n); //sorting
    int sum = 2000000001, L = 0, R =n-1, ans1=0, ans2=0;
    
    while(L<R){ //원소가 하나남은 경우는 포함되지 않으므로 '<'
        if(sum > abs(a[L] + a[R])){ //min value
            sum = abs(a[L] + a[R]); //update
            ans1 = L;
            ans2 = R;
        }
        a[L]+a[R] > 0 ? R-- : L++;
    }
    
    cout << a[ans1] << ' ' << a[ans2] << '\n';
    
}

int main(){
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    input();
    solution();
    return 0;
}
